src/backend/common/MemoryManager.hpp

```diff
@@ -236,71 +250,75 @@ class MemoryManager
 
-        // Pointer not found in locked map
-        if (iter == current.locked_map.end()) {
-            // Probably came from user, just free it
-            this->nativeFree(ptr);
-            return;
-        }
+            // Pointer not found in locked map
+            if (iter == current.locked_map.end()) {
+                // Probably came from user, just free it
+                freed_ptr.reset(ptr);
+                return;
+            }
 
-        if (user_unlock) {
-            (iter->second).user_lock = false;
-        } else {
-            (iter->second).manager_lock = false;
-        }
+            if (user_unlock) {
+                (iter->second).user_lock = false;
+            } else {
+                (iter->second).manager_lock = false;
+            }
 
-        // Return early if either one is locked
-        if ((iter->second).user_lock || (iter->second).manager_lock) return;
+            // Return early if either one is locked
+            if ((iter->second).user_lock || (iter->second).manager_lock) return;
 
-        size_t bytes = iter->second.bytes;
-        current.lock_bytes -= iter->second.bytes;
-        current.lock_buffers--;
+            size_t bytes = iter->second.bytes;
+            current.lock_bytes -= iter->second.bytes;
+            current.lock_buffers--;
 
-        if (this->debug_mode) {
-            // Just free memory in debug mode
-            if ((iter->second).bytes > 0) {
-                this->nativeFree(iter->first);
-                current.total_buffers--;
-                current.total_bytes -= iter->second.bytes;
-            }
-        } else {
-            // In regular mode, move buffer to free map
-            free_iter fiter = current.free_map.find(bytes);
-            if (fiter != current.free_map.end()) {
-                // If found, push back
-                fiter->second.push_back(ptr);
+            if (this->debug_mode) {
+                // Just free memory in debug mode
+                if ((iter->second).bytes > 0) {
+                    freed_ptr.reset(iter->first);
+                    current.total_buffers--;
+                    current.total_bytes -= iter->second.bytes;
+                }
             } else {
-                // If not found, create new vector for this size
-                std::vector<void *> ptrs;
-                ptrs.push_back(ptr);
-                current.free_map[bytes] = ptrs;
+                // In regular mode, move buffer to free map
+                free_iter fiter = current.free_map.find(bytes);
+                if (fiter != current.free_map.end()) {
+                    // If found, push back
+                    fiter->second.push_back(ptr);
+                } else {
+                    // If not found, create new vector for this size
+                    std::vector<void *> ptrs;
+                    ptrs.push_back(ptr);
+                    current.free_map[bytes] = ptrs;
+                }
```

Comment:
I believe this whole "regular mode" branch (L291-L301) can be replaced with this line:
```c++
current.free_map[bytes].push_back(ptr);
```
because [operator[]](http://en.cppreference.com/w/cpp/container/unordered_map/operator_at) on mapping containers constructs the vector for you if the key does not exist.

Reply:
I agree with this ^ I was not aware of this quirk of C++ standard containers when I wrote this code.